File: experiment-runner/examples/greenlab-profiling/codes/gpt/fasta.js
const IM = 139968;
const IA = 3877;
const IC = 29573;
let LAST = 42;

const ALU =
  "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAA";

const IUB = [
  ["a", 0.27],
  ["c", 0.12],
  ["g", 0.12],
  ["t", 0.27],
  ["B", 0.02],
  ["D", 0.02],
  ["H", 0.02],
  ["K", 0.02],
  ["M", 0.02],
  ["N", 0.02],
  ["R", 0.02],
  ["S", 0.02],
  ["V", 0.02],
  ["W", 0.02],
  ["Y", 0.02],
];

const HomoSapiens = [
  ["a", 0.3029549426680],
  ["c", 0.1979883004921],
  ["g", 0.1975473066391],
  ["t", 0.3015094502008],
];

function toCumulative(arr) {
  let cumul = 0.0;
  for (let i = 0; i < arr.length; i++) {
    cumul += arr[i][1];
    arr[i][1] = cumul;
  }
}

function fastaRandom(n, arr) {
  toCumulative(arr);
  let ret = "";
  for (let i = 0; i < n; i++) {
    let r = genRandom();
    for (let j = 0; j < arr.length; j++) {
      if (r < arr[j][1]) {
        ret += arr[j][0];
        break;
      }
    }
    if (ret.length >= 60) {
      console.log(ret.substring(0, 60));
      ret = ret.substring(60);
    }
  }
  if (ret.length > 0) console.log(ret);
}

function fastaRepeat(n, seq) {
  let ret = "";
  let seqi = 0;
  for (let i = 0; i < n; i++) {
    ret += seq[seqi];
    seqi++;
    if (seqi == seq.length) seqi = 0;
    if (ret.length >= 60) {
      console.log(ret.substring(0, 60));
      ret = ret.substring(60);
    }
  }
  if (ret.length > 0) console.log(ret);
}

function genRandom(max = 1) {
  LAST = (LAST * IA + IC) % IM;
  return (max * LAST) / IM;
}

function main(n) {
  console.log(">ONE Homo sapiens alu");
  fastaRepeat(n * 2, ALU);
  console.log(">TWO IUB ambiguity codes");
  fastaRandom(n * 3, IUB);
  console.log(">THREE Homo sapiens frequency");
  fastaRandom(n * 5, HomoSapiens);
}

main(1000);

File: experiment-runner/examples/greenlab-profiling/codes/gpt/fasta.cpp
#include <algorithm>
#include <array>
#include <vector>
#include <iostream>
#include <numeric>
#include <functional>

struct IUB
{
   float p;
   char c;
};

const std::string alu =
{
  "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
   "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
   "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
   "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
   "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
   "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
   "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"
};

std::array<IUB,15> iub =
{{
   { 0.27f, 'a' },
   { 0.12f, 'c' },
   { 0.12f, 'g' },
   { 0.27f, 't' },
   { 0.02f, 'B' },
   { 0.02f, 'D' },
   { 0.02f, 'H' },
   { 0.02f, 'K' },
   { 0.02f, 'M' },
   { 0.02f, 'N' },
   { 0.02f, 'R' },
   { 0.02f, 'S' },
   { 0.02f, 'V' },
   { 0.02f, 'W' },
   { 0.02f, 'Y' }
}};

std::array<IUB, 4> homosapiens =
{{
   { 0.3029549426680f, 'a' },
   { 0.1979883004921f, 'c' },
   { 0.1975473066391f, 'g' },
   { 0.3015094502008f, 't' }
}};

const int IM = 139968;
const float IM_RECIPROCAL = 1.0f / IM;

uint32_t gen_random()
{
   static const int IA = 3877, IC = 29573;
   static int last = 42;
   last = (last * IA + IC) % IM;
   return last;
}

char convert_trivial(char c)
{
   return c;
}

template<class iterator_type>
char make_repeat_generator(iterator_type& current, iterator_type first, iterator_type last)
{
   if (current == last)
      current = first;
   return *current++;
}

template<class iterator_type>
char convert_random(uint32_t random, iterator_type begin, iterator_type end)
{
   const float p = random * IM_RECIPROCAL;
   auto result = std::find_if(begin, end, [p] (IUB i) { return p <= i.p; });
   return result->c;
}

char convert_IUB(uint32_t random)
{
   return convert_random(random, iub.begin(), iub.end());
}

char convert_homosapiens(uint32_t random)
{
   return convert_random(random, homosapiens.begin(), homosapiens.end());
}

template<class iterator_type>
void make_cumulative(iterator_type first, iterator_type last)
{
   std::partial_sum(first, last, first,
                [] (IUB l, IUB r) -> IUB { r.p += l.p; return r; });
}

const size_t CHARS_PER_LINE = 60;

template <class generator_type, class converter_type >
void make(const char* desc, int n, generator_type generator, converter_type converter) {
   std::cout << '>' << desc << '\n';
   
   int col = 0;
   for(int i = 0; i < n; ++i) {
      char output = converter(generator());
      std::cout << output;
      if(++col >= CHARS_PER_LINE) {
         std::cout << '\n';
         col = 0;
      }
   }
   if(col != 0) std::cout << '\n';
}

int main(int argc, char *argv[])
{
   int n = 1000;
   if (argc < 2 || (n = std::atoi(argv[1])) <= 0) {
      std::cerr << "usage: " << argv[0] << " length\n";
      return 1;
   }
   
   make_cumulative(iub.begin(), iub.end());
   make_cumulative(homosapiens.begin(), homosapiens.end());

   auto alu_iter = alu.begin();
   make("ONE Homo sapiens alu", n * 2,
       [&]() { return make_repeat_generator(alu_iter, alu.begin(), alu.end()); },
       convert_trivial );
   
   make("TWO IUB ambiguity codes", n * 3,
       [&]() { return gen_random(); },
       convert_IUB );
   
   make("THREE Homo sapiens frequency", n * 5,
       [&]() { return gen_random(); },
       convert_homosapiens );
   
   return 0;
}
